lib/public-stats.js

```diff
@@ -104,37 +106,134 @@ const updateIndexerQueryStats = async (pgClient, committees) => {
 /**
  * @param {pg.Client} pgClient
  * @param {Iterable<Committee>} committees
+ * @param {(minerId: string, cid: string) => (string[] | undefined)} findDealClients
  */
-const updateDailyDealsStats = async (pgClient, committees) => {
-  let total = 0
-  let indexed = 0
-  let retrievable = 0
+const updateDailyDealsStats = async (pgClient, committees, findDealClients) => {
+  /** @type {Map<string, Map<string, {
+   * tested: number;
+   * index_majority_found: number;
+   * retrieval_majority_found: number;
+   * indexed:  number;
+   * indexed_http: number;
+   * retrievable: number;
+   * }>>} */
+  const minerClientDealStats = new Map()
   for (const c of committees) {
-    total++
+    const { minerId, cid } = c.retrievalTask
+    const clients = findDealClients(minerId, cid)
+    if (!clients || !clients.length) {
+      console.warn(`Invalid retrieval task (${minerId}, ${cid}): no deal clients found. Excluding the task from daily per-deal stats.`)
+      Sentry.captureException(new Error('Invalid retrieval task: no deal clients found.'), {
+        extra: {
+          minerId,
+          cid
+        }
+      })
+      continue
+    }
 
-    const evaluation = c.evaluation
-    if (!evaluation) continue
-    if (evaluation.indexerResult === 'OK' || evaluation.indexerResult === 'HTTP_NOT_ADVERTISED') {
-      indexed++
+    let clientDealStats = minerClientDealStats.get(minerId)
+    if (!clientDealStats) {
+      clientDealStats = new Map()
+      minerClientDealStats.set(minerId, clientDealStats)
     }
-    if (evaluation.retrievalResult === 'OK') {
-      retrievable++
+
+    for (const clientId of clients) {
+      let stats = clientDealStats.get(clientId)
+      if (!stats) {
+        stats = {
+          tested: 0,
+          index_majority_found: 0,
+          retrieval_majority_found: 0,
+          indexed: 0,
+          indexed_http: 0,
+          retrievable: 0
+        }
+        clientDealStats.set(clientId, stats)
+      }
+
+      stats.tested++
+
+      const evaluation = c.evaluation
+      if (!evaluation) continue
+
+      if (evaluation.indexerResult !== 'COMMITTEE_TOO_SMALL' && evaluation.indexerResult !== 'MAJORITY_NOT_FOUND') {
+        stats.index_majority_found++
+      }
+
+      if (evaluation.indexerResult === 'OK' || evaluation.indexerResult === 'HTTP_NOT_ADVERTISED') {
+        stats.indexed++
+      }
+
+      if (evaluation.indexerResult === 'OK') {
+        stats.indexed_http++
+      }
+
+      if (evaluation.retrievalResult !== 'COMMITTEE_TOO_SMALL' && evaluation.indexerResult !== 'MAJORITY_NOT_FOUND') {
```

Comment:
BUG! The second condition is incorrectly testing `indexerResult` instead of `retrievalResult`.

Reply:
I fixed this in [d675263](https://github.com/filecoin-station/spark-evaluate/pull/336/commits/d67526348fd5689d734b649fafa2260f1a26bfff)